Simplified/Utilities/Authentication/NYPLUserAccountFrontEndValidation.swift

```diff
@@ -0,0 +1,89 @@
+//
+//  NYPLUserAccountFrontEndValidation.swift
+//  SimplyE
+//
+//  Created by Jacek Szyja on 26/05/2020.
+//  Copyright © 2020 NYPL Labs. All rights reserved.
+//
+
+import UIKit
+
+/**
+ Protocol that represents the input sources / UI requirements for performing
+ front-end validation.
+ */
+@objc
+protocol NYPLUserAccountInputProvider {
+  var usernameTextField: UITextField! { get set }
+  var PINTextField: UITextField! { get set }
+}
+
+@objcMembers class NYPLUserAccountFrontEndValidation: NSObject {
+  let userInputProvider: NYPLUserAccountInputProvider
+  let account: Account
+  private weak var businessLogic: NYPLSignInBusinessLogic?
+
+  init(account: Account, businessLogic: NYPLSignInBusinessLogic?, inputProvider: NYPLUserAccountInputProvider) {
+    self.userInputProvider = inputProvider
+    self.account = account
+    self.businessLogic = businessLogic
+  }
+}
+
+extension NYPLUserAccountFrontEndValidation: UITextFieldDelegate {
+  func textFieldShouldBeginEditing(_ textField: UITextField) -> Bool {
+    return !(businessLogic?.userAccount.hasBarcodeAndPIN() ?? false)
+  }
+
+  func textField(_ textField: UITextField, shouldChangeCharactersIn range: NSRange, replacementString string: String) -> Bool {
+    guard string.canBeConverted(to: .ascii) else { return false }
+
+    if textField == userInputProvider.usernameTextField,
+      businessLogic?.selectedAuthentication?.patronIDKeyboard != .email {
+
+      // Barcodes are numeric and usernames are alphanumeric including punctuation
+      let allowedCharacters = CharacterSet.alphanumerics.union(.punctuationCharacters)
+      let bannedCharacters = allowedCharacters.inverted
+
+      guard string.rangeOfCharacter(from: bannedCharacters) == nil else { return false }
+
+      if let text = textField.text,
+        let textRange = Range(range, in: text) {
+
+        let updatedText = text.replacingCharacters(in: textRange, with: string)
+        guard updatedText.count <= 25 else { return false }
```

Comment:
can you add a comment that 25 is a server requirement?

Reply:
must have skipped that, added it back